refactor: small optimization for IsHex method

shargon opened the pull request:
In case the condition is `false`, every character is checked first. This operation is not necessary.

If you call the method with a truncated hex string, you will loose performance. The last check will be done, so is better to do it the first one

shargon commented on src/util/strencodings.cpp:
```diff
@@ -59,12 +59,13 @@ signed char HexDigit(char c)
 
 bool IsHex(const std::string& str)
 {
+    if (str.size() <= 0 || str.size()%2 != 0) return false;
```
Choose your favorite:
```suggestion
    if (str.size() == 0 || str.size()%2 != 0) return false;
```

fanquake commented:
Please fix your commit messages and remove the merge commit.

As this is an `optimization`, please also provide benchmarks.

promag reviewed (commented):
This change supposedly optimizes the false case, I don't think it matters.

jcliff commented:
+1 to @fanquake's recommendation for squashing the commits and fixing the commit message to be clearer. 

@fanquake, @promag, This change clearly avoids computation in some cases without adding complexity.

shargon committed:
Small optimization for IsHex method

In the case of false, we check every character first, not needed
Small optimization for IsHex method [size==0]

promag commented:
> This change clearly avoids computation in some cases without adding complexity.

@jcliff and in other cases adds more "computation". I suspect a benchmark would give similar results, and in that case why bother?

laanwj commented:
Is there any user-visible (or, network-visible) behavior that this would speed up?

Thank you for your first contribution, but we're kind of strict in this project as to requiring a strong rationale for every change. It takes effort to review a change, and there's always a risk.

fqlx commented on src/util/strencodings.cpp:
```diff
@@ -59,12 +59,13 @@ signed char HexDigit(char c)
 
 bool IsHex(const std::string& str)
 {
+    if (str.size() == 0 || str.size()%2 != 0) return false;
```
Please add spaces around the modulus operator.

shargon replied:
Before is like this

laanwj commented:
Closing this, sorry, there is no agreement to do this
